source/common/http/header_map_impl.h

```diff
@@ -61,12 +57,11 @@ public:
  */
 class HeaderMapImpl : public virtual HeaderMap, NonCopyable {
 public:
-  HeaderMapImpl();
   // The following "constructors" call virtual functions during construction and must use the
   // static factory pattern.
-  static void copyFrom(HeaderMapImpl& lhs, const HeaderMap& rhs);
+  static void copyFrom(HeaderMap& lhs, const HeaderMap& rhs);
```

Comment:
Is the reason you did it like this so that you could share the base-class implementation?

How about putting the guts into `copyFromHelper` in the `protected` session and having type-specific `copyFrom` inline calls to that, so we could avoid copying between types?

Reply:
I honestly didn't think about it that much. There are good points. Let me see if I can do better here.

Reply:
I looked at this and this is not so easy due to some of the on-purpose conversions we do for gRPC. I'm going to add a TODO and punt on this for PR just so I can finish this and then we can look at this in a follow up. Let me know how that sounds.